File: sdk-typescript/packages/captions/src/captions.ts
/**
 * WAVE SDK - Captions API
 *
 * Generate, manage, and translate captions for video content.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
} from '@wave-av/core';
import type { CaptionFormat, CaptionTrack, CaptionCue, GenerateCaptionsRequest, UploadCaptionsRequest, UpdateCaptionsRequest, TranslateCaptionsRequest, BurnInCaptionsRequest, BurnInJob, ListCaptionsParams } from './captions-types';
export type { CaptionStatus, CaptionFormat, CaptionTrack, CaptionCue, CaptionWord, CaptionStyle, GenerateCaptionsRequest, UploadCaptionsRequest, UpdateCaptionsRequest, TranslateCaptionsRequest, BurnInCaptionsRequest, BurnInJob, ListCaptionsParams } from './captions-types';

// Types

/**
 * Caption status
 */

/**
 * Caption format
 */

/**
 * Caption track
 */

/**
 * Caption cue (single caption segment)
 */

/**
 * Word-level timing
 */

/**
 * Caption styling
 */

/**
 * Generate captions request
 */

/**
 * Upload captions request
 */

/**
 * Update captions request
 */

/**
 * Translate captions request
 */

/**
 * Burn-in captions request
 */

/**
 * Burn-in job
 */

/**
 * List captions params
 */

// Captions API

/**
 * Captions API client
 */
export class CaptionsAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/captions';

  constructor(client: WaveClient) {
    this.client = client;
  }

  /**
   * Generate captions using AI
   *
   * Requires: captions:generate permission
   */
  async generate(request: GenerateCaptionsRequest): Promise<CaptionTrack> {
    return this.client.post<CaptionTrack>(`${this.basePath}/generate`, request);
  }

  /**
   * Upload existing captions
   *
   * Requires: captions:create permission
   */
  async upload(request: UploadCaptionsRequest): Promise<CaptionTrack> {
    return this.client.post<CaptionTrack>(`${this.basePath}/upload`, request);
  }

  /**
   * Get a caption track by ID
   *
   * Requires: captions:read permission
   */
  async get(trackId: string): Promise<CaptionTrack> {
    return this.client.get<CaptionTrack>(`${this.basePath}/${trackId}`);
  }

  /**
   * Update a caption track
   *
   * Requires: captions:update permission
   */
  async update(trackId: string, request: UpdateCaptionsRequest): Promise<CaptionTrack> {
    return this.client.patch<CaptionTrack>(`${this.basePath}/${trackId}`, request);
  }

  /**
   * Remove a caption track
   *
   * Requires: captions:remove permission (server-side RBAC enforced)
   */
  async remove(trackId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${trackId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * List caption tracks
   *
   * Requires: captions:read permission
   */
  async list(params?: ListCaptionsParams): Promise<PaginatedResponse<CaptionTrack>> {
    return this.client.get<PaginatedResponse<CaptionTrack>>(this.basePath, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }

  /**
   * Get caption tracks for a specific media
   *
   * Requires: captions:read permission
   */
  async getForMedia(
    mediaId: string,
    mediaType: 'video' | 'audio' | 'stream'
  ): Promise<CaptionTrack[]> {
    const result = await this.list({ media_id: mediaId, media_type: mediaType });
    return result.data;
  }

  /**
   * Get caption cues (segments)
   *
   * Requires: captions:read permission
   */
  async getCues(
    trackId: string,
    params?: PaginationParams & { start_time?: number; end_time?: number }
  ): Promise<PaginatedResponse<CaptionCue>> {
    return this.client.get<PaginatedResponse<CaptionCue>>(
      `${this.basePath}/${trackId}/cues`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Update a caption cue
   *
   * Requires: captions:update permission
   */
  async updateCue(
    trackId: string,
    cueId: string,
    updates: Partial<Pick<CaptionCue, 'text' | 'start_time' | 'end_time' | 'speaker' | 'style'>>
  ): Promise<CaptionCue> {
    return this.client.patch<CaptionCue>(
      `${this.basePath}/${trackId}/cues/${cueId}`,
      updates
    );
  }

  /**
   * Add a new caption cue
   *
   * Requires: captions:update permission
   */
  async addCue(
    trackId: string,
    cue: Omit<CaptionCue, 'id' | 'confidence' | 'words'>
  ): Promise<CaptionCue> {
    return this.client.post<CaptionCue>(
      `${this.basePath}/${trackId}/cues`,
      cue
    );
  }

  /**
   * Remove a caption cue
   *
   * Requires: captions:update permission (server-side RBAC enforced)
   */
  async removeCue(trackId: string, cueId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${trackId}/cues/${cueId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Bulk update cues
   *
   * Requires: captions:update permission
   */
  async bulkUpdateCues(
    trackId: string,
    updates: Array<{ id: string; text?: string; start_time?: number; end_time?: number }>
  ): Promise<{ updated: number }> {
    return this.client.post(`${this.basePath}/${trackId}/cues/bulk`, { updates });
  }

  /**
   * Translate a caption track to another language
   *
   * Requires: captions:translate permission
   */
  async translate(
    trackId: string,
    request: TranslateCaptionsRequest
  ): Promise<CaptionTrack> {
    return this.client.post<CaptionTrack>(
      `${this.basePath}/${trackId}/translate`,
      request
    );
  }

  /**
   * Export captions in a specific format
   *
   * Requires: captions:read permission
   */
  async exportFormat(
    trackId: string,
    format: CaptionFormat
  ): Promise<{ url: string; expires_at: string }> {
    return this.client.get(`${this.basePath}/${trackId}/export`, {
      params: { format } as unknown as Record<string, string | number | boolean | undefined>,
    });
  }

  /**
   * Get captions as plain text
   *
   * Requires: captions:read permission
   */
  async getText(trackId: string): Promise<string> {
    const result = await this.client.get<{ text: string }>(
      `${this.basePath}/${trackId}/text`
    );
    return result.text;
  }

  /**
   * Burn captions into video
   *
   * Requires: captions:burnin permission
   */
  async burnIn(request: BurnInCaptionsRequest): Promise<BurnInJob> {
    return this.client.post<BurnInJob>(`${this.basePath}/burn-in`, request);
  }

  /**
   * Get burn-in job status
   *
   * Requires: captions:read permission
   */
  async getBurnInJob(jobId: string): Promise<BurnInJob> {
    return this.client.get<BurnInJob>(`${this.basePath}/burn-in/${jobId}`);
  }

  /**
   * Wait for burn-in to complete
   */
  async waitForBurnIn(
    jobId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (job: BurnInJob) => void;
    }
  ): Promise<BurnInJob> {
    const pollInterval = options?.pollInterval || 3000;
    const timeout = options?.timeout || 1800000; // 30 minutes
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const job = await this.getBurnInJob(jobId);

      if (options?.onProgress) {
        options.onProgress(job);
      }

      if (job.status === 'ready') {
        return job;
      }

      if (job.status === 'failed') {
        throw new Error(`Burn-in failed: ${job.error || 'Unknown error'}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Burn-in timed out after ${timeout}ms`);
  }

  /**
   * Wait for caption generation to complete
   */
  async waitForReady(
    trackId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (track: CaptionTrack) => void;
    }
  ): Promise<CaptionTrack> {
    const pollInterval = options?.pollInterval || 2000;
    const timeout = options?.timeout || 600000; // 10 minutes
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const track = await this.get(trackId);

      if (options?.onProgress) {
        options.onProgress(track);
      }

      if (track.status === 'ready') {
        return track;
      }

      if (track.status === 'failed') {
        throw new Error(`Caption generation failed: ${track.error || 'Unknown error'}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Caption generation timed out after ${timeout}ms`);
  }

  /**
   * Get supported languages
   *
   * Requires: captions:read permission
   */
  async getSupportedLanguages(): Promise<
    Array<{
      code: string;
      name: string;
      native_name: string;
      supports_generation: boolean;
      supports_translation: boolean;
    }>
  > {
    return this.client.get(`${this.basePath}/languages`);
  }

  /**
   * Detect language from audio
   *
   * Requires: captions:generate permission
   */
  async detectLanguage(
    mediaId: string,
    mediaType: 'video' | 'audio' | 'stream'
  ): Promise<{
    detected_language: string;
    confidence: number;
    alternatives: Array<{ language: string; confidence: number }>;
  }> {
    return this.client.post(`${this.basePath}/detect-language`, {
      media_id: mediaId,
      media_type: mediaType,
    });
  }
}

/**
 * Create a Captions API instance
 */
export function createCaptionsAPI(client: WaveClient): CaptionsAPI {
  return new CaptionsAPI(client);
}

File: sdk-python/wave_sdk/captions.py
"""WAVE SDK - Captions API. Auto-generate, translate, burn-in, and manage captions."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class CaptionTrack(BaseModel):
    id: str; asset_id: str; language: str; status: str; format: str; cue_count: int = 0; download_url: str | None = None; created_at: str; updated_at: str

class CaptionCue(BaseModel):
    id: str; start_time: float; end_time: float; text: str; speaker_id: str | None = None

class BurnInJob(BaseModel):
    id: str; caption_id: str; status: str; progress_percent: int = 0; output_url: str | None = None; error: str | None = None; created_at: str

class CaptionsAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/captions"
    def generate(self, media_id: str, media_type: str = "video", language: str = "en", **kwargs: Any) -> CaptionTrack: return CaptionTrack(**self._client.post(self._base, json={"media_id": media_id, "media_type": media_type, "language": language, **kwargs}))
    def get(self, caption_id: str) -> CaptionTrack: return CaptionTrack(**self._client.get(f"{self._base}/{caption_id}"))
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def update(self, caption_id: str, **kwargs: Any) -> CaptionTrack: return CaptionTrack(**self._client.patch(f"{self._base}/{caption_id}", json=kwargs))
    def remove(self, caption_id: str) -> None: self._client.delete(f"{self._base}/{caption_id}")
    def upload(self, media_id: str, language: str, format: str, content: str, **kwargs: Any) -> CaptionTrack: return CaptionTrack(**self._client.post(f"{self._base}/upload", json={"media_id": media_id, "language": language, "format": format, "content": content, **kwargs}))
    def translate(self, caption_id: str, target_language: str) -> CaptionTrack: return CaptionTrack(**self._client.post(f"{self._base}/{caption_id}/translate", json={"target_language": target_language}))
    def get_cues(self, caption_id: str) -> list[CaptionCue]: return [CaptionCue(**c) for c in self._client.get(f"{self._base}/{caption_id}/cues")]
    def add_cue(self, caption_id: str, start_time: float, end_time: float, text: str, **kwargs: Any) -> CaptionCue: return CaptionCue(**self._client.post(f"{self._base}/{caption_id}/cues", json={"start_time": start_time, "end_time": end_time, "text": text, **kwargs}))
    def update_cue(self, caption_id: str, cue_id: str, **kwargs: Any) -> CaptionCue: return CaptionCue(**self._client.patch(f"{self._base}/{caption_id}/cues/{cue_id}", json=kwargs))
    def remove_cue(self, caption_id: str, cue_id: str) -> None: self._client.delete(f"{self._base}/{caption_id}/cues/{cue_id}")
    def bulk_update_cues(self, caption_id: str, cues: list[dict]) -> list[CaptionCue]: return [CaptionCue(**c) for c in self._client.put(f"{self._base}/{caption_id}/cues", json={"cues": cues})]
    def export_format(self, caption_id: str, format: str = "srt") -> dict: return self._client.get(f"{self._base}/{caption_id}/export", params={"format": format})
    def get_text(self, caption_id: str, include_speakers: bool = False) -> dict: return self._client.get(f"{self._base}/{caption_id}/text", params={"include_speakers": include_speakers})
    def detect_language(self, media_id: str) -> dict: return self._client.post(f"{self._base}/detect-language", json={"media_id": media_id})
    def get_supported_languages(self) -> list[dict]: return self._client.get(f"{self._base}/languages")
    def get_for_media(self, media_id: str, media_type: str = "video") -> list[CaptionTrack]: return [CaptionTrack(**t) for t in self._client.get(f"{self._base}/media/{media_id}", params={"media_type": media_type})]
    def burn_in(self, caption_id: str, **kwargs: Any) -> BurnInJob: return BurnInJob(**self._client.post(f"{self._base}/{caption_id}/burn-in", json=kwargs))
    def get_burn_in_job(self, job_id: str) -> BurnInJob: return BurnInJob(**self._client.get(f"{self._base}/burn-in/{job_id}"))
    def wait_for_ready(self, caption_id: str, poll_interval: float = 2.0, timeout: float = 300.0) -> CaptionTrack:
        start = time.time()
        while time.time() - start < timeout:
            t = self.get(caption_id)
            if t.status == "ready": return t
            if t.status == "failed": raise RuntimeError("Caption generation failed")
            time.sleep(poll_interval)
        raise TimeoutError(f"Caption timed out after {timeout}s")
    def wait_for_burn_in(self, job_id: str, poll_interval: float = 3.0, timeout: float = 600.0) -> BurnInJob:
        start = time.time()
        while time.time() - start < timeout:
            j = self.get_burn_in_job(job_id)
            if j.status == "ready": return j
            if j.status == "failed": raise RuntimeError(f"Burn-in failed: {j.error}")
            time.sleep(poll_interval)
        raise TimeoutError(f"Burn-in timed out after {timeout}s")
